xarray/backends/zarr.py

```diff
@@ -0,0 +1,172 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+import functools
+import warnings
+from itertools import product
+from collections import MutableMapping
+
+from .. import Variable
+from ..core import indexing
+from ..core.utils import FrozenOrderedDict, close_on_error, HiddenKeyDict
+from ..core.pycompat import iteritems, bytes_type, unicode_type, OrderedDict
+
+from .common import (WritableCFDataStore, AbstractWritableDataStore,
+                     DataStorePickleMixin)
+
+
+
+
+# most of the other stores have some kind of wrapper class like
+# class BaseNetCDF4Array(NdimSizeLenMixin, DunderArrayMixin):
+# class H5NetCDFArrayWrapper(BaseNetCDF4Array):
+# class NioArrayWrapper(NdimSizeLenMixin, DunderArrayMixin):
+# we problaby need something like this
+
+# the first question is whether it should be based on BaseNetCDF4Array or
+# NdimSizeLenMixing?
+
+# or maybe we don't need wrappers at all? probably not true
+
+
+# also most have a custom opener
+
+# keyword args for zarr.group
+# store=None, overwrite=False, chunk_store=None, synchronizer=None, path=None
+# the group name is called "path" in the zarr lexicon
+
+def _open_zarr_group(store, overwrite, chunk_store, synchronizer, path):
+    import zarr
+    zarr_group = zarr.group(store=store, overwrite=overwrite,
+            chunk_store=chunk_store, synchronizer=synchronizer, path=path)
+    return zarr_group
+
+
+def _dask_chunks_to_zarr_chunks(chunks):
+    # zarr chunks needs to be uniform for each array
+    # http://zarr.readthedocs.io/en/latest/spec/v1.html#chunks
+    # dask chunks can be variable sized
+    # http://dask.pydata.org/en/latest/array-design.html#chunks
+    # this function dask chunks syntax to zarr chunks
+    if chunks is None:
+        return chunks
+
+    all_chunks = product(*chunks)
+    first_chunk = all_chunks.next()
+    for this_chunk in all_chunks:
+        if not (this_chunk == first_chunk):
+            raise ValueError("zarr requires uniform chunk sizes, found %s" %
+                             repr(chunks))
+    return first_chunk
+
+
+def _get_zarr_dims_and_attrs(zarr_obj, dimension_key):
+    # Zarr arrays do not have dimenions. To get around this problem, we add
+    # an attribute that specifies the dimension. We have to hide this attribute
+    # when we send the attributes to the user.
+    # zarr_obj can be either a zarr group or zarr array
+    dimensions = zarr_obj.attrs.get(dimension_key)
+    attributes = HiddenKeyDict(zarr_obj.attrs, dimension_key)
+    return dimensions, attributes
+
+
+class ZarrStore(AbstractWritableDataStore, DataStorePickleMixin):
+    """Store for reading and writing data via zarr
+    """
+
+    # need some special secret attributes to tell us the dimensions
+    _dimension_key = '_XARRAY_DIMENSIONS'
```

Comment:
nit: should be `_DIMENSION_KEY` since it's a constant.

Also: maybe better to pick something more generic for the constant value, perhaps `'_ARRAY_DIMENSIONS'`?